### PR TITLE
Add Binnie's hive generation to the Forestry HiveRegistry

### PR DESCRIPTION
**Pull request by temp1011**
Adds Binnie hive generation to forestry `HiveRegistry` so that binnie hives will obey forestry configs (eg - for dimension blacklisting.

Still TODO:
~- figure out why this seems to get stuck sometimes~
~- expose getting the instance of `HiveRegistry` in Forestry~
~- Make this faster  (worldgen is very slow).~

Hopefully it is possible to do this without it being too slow. I'm sure there's a lot that can be done to help that. I'm probably misusing `IHivegen` at the moment.

### CODE REVIEW

File: extrabees/src/main/java/binnie/extrabees/worldgen/WorldGenHiveMarble.java [hunk 1]
```diff
@@ -47,4 +49,47 @@ public boolean generate(World world, Random rand, BlockPos position) {
...
+			return null;
+		}
+
+		final BlockPos.MutableBlockPos pos = new BlockPos.MutableBlockPos(topPos);
+
+		while (pos.getY() > 0) {
+			if (isValidLocation(world, pos)) {
+				return pos;
+			}
+			pos.add(0, -1, 0);
```
Comment (mezz) [inline extrabees/src/main/java/binnie/extrabees/worldgen/WorldGenHiveMarble.java]: The strategy here will check every block from the surface down to 0.
A more efficient strategy would be to get the  `topPos` and try a few random locations between 0 and `topPos`.
Only try a few locations before giving up.

**Review by mezz — CHANGES_REQUESTED** [review]

**Comment by temp1011** [comment]
Still needs tweaks to get worldgen to previous levels. Also, the rate registered with forestry should depend on the config settings for extrabees.

File: extrabees/src/main/java/binnie/extrabees/worldgen/BinnieHiveDescription.java [hunk 2]
```diff
@@ -22,42 +24,41 @@
...
 
 	@Override
 	public IBlockState getBlockState() {
-		return hiveState;
+		return ExtraBees.hive.getDefaultState().withProperty(BlockExtraBeeHives.HIVE_TYPE, hiveType);
 	}
 
 	@Override
 	public boolean isGoodBiome(Biome biome) {
+		if (hiveType == EnumHiveType.NETHER) {
```
Comment (KorDum) [inline extrabees/src/main/java/binnie/extrabees/worldgen/BinnieHiveDescription.java]: You can simplify this
```java
return hiveType != EnumHiveType.NETHER
    || BiomeDictionary.hasType(biome, BiomeDictionary.Type.NETHER);
```

Comment (temp1011) [reply extrabees/src/main/java/binnie/extrabees/worldgen/BinnieHiveDescription.java]: Thanks

**Comment by temp1011** [comment]
Current gen with hive debug on.

**Comment by temp1011** [comment]
I'm still not sure if the rates are right. Also, I think marble hives were designed for Red Power 2 marble so don't generate as frequently on chisel marble. I will include some issues I think this resolves:
Closes https://github.com/ForestryMC/Binnie/issues/459
Closes https://github.com/ForestryMC/Binnie/issues/424
Closes https://github.com/ForestryMC/Binnie/issues/393 (as far as I am aware no trees generate in the nether and Forestry does not have cascading issues)
Closes https://github.com/ForestryMC/Binnie/issues/301 (similarly I don't think trees generate in the deep dark)
Note there is still an issue open for cascading worldgen in https://github.com/ForestryMC/Binnie/issues/315 which needs the forge PR to solve.

**Comment by mezz** [comment]
Whenever I've done tweaks to hive generation, I add some debug logging for how many of each hive are placed.
Start a new world and fly forward for a specific distance and see how many of each hive were placed. Recreate the world with the same seed and do the same thing with your changes applied, and tweak the values until they are close.

**Comment by AnrDaemon** [comment]
> as far as I am aware no trees generate in the nether

[You would be surprized…](https://minecraft.curseforge.com/projects/natura)

**Comment by Nedelosk** [comment]
@AnrDaemon He talks about trees from extra trees.

**Comment by Nedelosk** [comment]
@temp1011 What's the current state of this PR ?

**Comment by temp1011** [comment]
I haven't looked at it in a while sorry. I need to test the numbers to make sure everything works the same. Not sure how good this is for performance as well.

**Comment by acanady** [comment]
Oh great though I just couldn't find them, turns out they weren't generating at all to begin with

**Comment by temp1011** [comment]
The bee effect coming off the hives is the wrong colour at the moment too (I get forest bees for rocky hives).

**Comment by temp1011** [comment]
I can hopefully get this done this Sunday. At the moment it's difficult for me to test it though.

**Comment by Nedelosk** [comment]
@temp1011 Any update ?

**Comment by temp1011** [comment]
I believe this needs https://github.com/ForestryMC/ForestryMC/pull/2204 to work well, but I haven't looked at this in a while.

**Comment by temp1011** [comment]
See new PR

**Pull request closed without merging**